<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>async</title>
</head>
<body>
<script>
    /**
     * async function(){
     *   await...
     *   return;
     * }
     *
     *  await 自动执行后面的Promise表达式（后面的表达式必须返回的式Promise对象）
     *  当Promise的异步操作执行完毕之后，函数才会继续向下执行。
     *  return ;可以通过.then()接受
     * */
    let p1 = function (){
        return new Promise((resolve,reject) => {
            setTimeout(() => {
                console.log(1);
                resolve("成功")
            },3000)
        });
    };
    let p2 = function (data){
        return new Promise((resolve,reject) => {
            setTimeout(() => {
                console.log(`${data}  这里是p2`);
                resolve(`已经全部完成`)
            },1000)
        });
    };
    /*
     当遇到await 自动执行关键字后面的Promise对象
     当Promise对象变为成功或失败 会自动继续执行下一个await后面的Promise
     * */
    async function fn(){
        // 只有在async函数里面才能使用await关键字
        //  Promise对象成功时返回数据 可以在await关键字前面定义一个变量来接收
        // 可以通过trycatch捕获异常
        try {
            let pp1 = await p1();
            let pp2 = await p2(pp1);
            // 这里的return出去的东西是在.then里面接收
            return pp2
        }catch(err){
            console.error(err)
        }
    }
    fn().then((data) => {
        console.log(data);
    });

</script>
</body>
</html>